### PR TITLE
Provide initial compatibility with PHP 8

### LINKED ISSUE #2222: Aggregation\Stage\Match class won't work with PHP8
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | no
| RFC         | no
| BC Break    | yes

#### Summary

Just a heads up if no one noticed it before, but [`Match`](https://github.com/doctrine/mongodb-odm/blob/master/lib/Doctrine/ODM/MongoDB/Aggregation/Stage/Match.php) will be a reserved word in PHP8 so that won't work with PHP8 anymore 🙂

### PR DESCRIPTION
**Pull request by alcaeus**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | Fixes #2222.

#### Summary

This PR makes the code base compatible with PHP 8. A few caveats:
* The `match` functionality in aggregation builder requires changing as `match` is now a reserved keyword. This PR includes changes from #2237.
* Tests are run on PHP 8, but we need to remove a few dev depdendencies as they block installing a PHP 8 compatible PHPUnit version.
* Until `ocramius/proxy-manager` is compatible with PHP 8 (see https://github.com/Ocramius/ProxyManager/pull/628) we need to run `composer install` on PHP 8 with `--ignore-platform-req=php`.
* Proxies and custom collections are not yet fully compatible with PHP 8 features such as union types. The former will be handled by ProxyManager but requires tests, while the latter will have to be taken care of in a separate pull request.

### CODE REVIEW

**Review by SenseException — APPROVED** [review]

**Comment by jcaillot** [comment]
Thank you!

**Comment by alcaeus** [comment]
Note: I'll hold off on merging this until I've finished migrating the builds to GitHub actions. Can't rely on travis-CI anymore.

**Comment by garak** [comment]
Any news on this? I see that CI is running on github actions already

**Comment by alcaeus** [comment]
I lost track of this but will revisit soon. Thanks for the reminder 👍

**Comment by malarzm** [comment]
I've opened https://github.com/doctrine/mongodb-odm/issues/2258 to track persistent collection changes

**Comment by malarzm** [comment]
I have merged `2.1.x` branch into the `master` recently so you'll need to rebase your branch.

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup/Match.php [hunk 1]
```diff
@@ -4,106 +4,9 @@
...
 namespace Doctrine\ODM\MongoDB\Aggregation\Stage\GraphLookup;
 
-use Doctrine\ODM\MongoDB\Aggregation\Builder;
-use Doctrine\ODM\MongoDB\Aggregation\Expr;
-use Doctrine\ODM\MongoDB\Aggregation\Stage\GraphLookup;
-use Doctrine\ODM\MongoDB\Aggregation\Stage\Match as BaseMatch;
-
-class Match extends BaseMatch
+/**
+ * @deprecated Deprecated in favour of the MatchStage class
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup/Match.php:8]: This does not follow our deprecation template

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup/Match.php [hunk 2]
```diff
@@ -4,106 +4,9 @@
...
+ */
+class Match extends MatchStage
 {
-    /** @var GraphLookup */
-    private $graphLookup;
-
-    public function __construct(Builder $builder, GraphLookup $graphLookup)
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup/Match.php:17]: I'd be nice to have ctor here and trigger an in-code deprecation

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/Match.php [hunk 3]
```diff
@@ -4,518 +4,9 @@
...
-use Doctrine\ODM\MongoDB\Query\Expr;
-use GeoJson\Geometry\Geometry;
-use function func_get_args;
-
 /**
- * Fluent interface for building aggregation pipelines.
+ * @deprecated Deprecated in favour of the MatchStage class
  */
-class Match extends Stage
+class Match extends MatchStage
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Stage/Match.php:10]: Same comments as above

File: lib/Doctrine/ODM/MongoDB/PersistentCollection/DefaultPersistentCollectionGenerator.php [hunk 4]
```diff
@@ -280,10 +271,8 @@ private function getParameterNamesForDecoratedCall(array $parameters) : array
             static function (ReflectionParameter $parameter) {
                 $name = '';
 
-                if (method_exists($parameter, 'isVariadic')) {
-                    if ($parameter->isVariadic()) {
-                        $name .= '...';
-                    }
+                if ($parameter->isVariadic()) {
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/PersistentCollection/DefaultPersistentCollectionGenerator.php:274]: Thanks for tidying this up a bit 👍

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php [hunk 5]
```diff
@@ -509,22 +509,22 @@ public static function queryProviderForComplexRefWithObjectValue() : Generator
     {
         yield 'Direct comparison' => [
             'expected' => ['complexRef.date' => new UTCDateTime('1590710400000')],
-            'query' => ['complexRef.date' => new DateTime('2020-05-29')],
+            'query' => ['complexRef.date' => DateTime::createFromFormat('U', '1590710400')],
         ];
 
         yield 'Operator with single value' => [
             'expected' => ['complexRef.date' => ['$ne' => new UTCDateTime('1590710400000')]],
-            'query' => ['complexRef.date' => ['$ne' => new DateTime('2020-05-29')]],
```
Comment (malarzm) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php:517]: what's the reason of this change?

**Review by malarzm — APPROVED** [review]

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php (hunk 5 above)
Comment (alcaeus) [reply tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php:517]: Timezones ;)

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup/Match.php (hunk 2 above)
Comment (alcaeus) [reply lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup/Match.php:17]: Good point, will add 👍

**Comment by alcaeus** [comment]
@malarzm I waited long enough to take care of proxies as well: we can now generate proxies with 7.4 language features on PHP 8 by switching to friendsofphp/proxy-manager-lts. Proxy support for PHP 8 features is still outstanding pending an upstream fix.

**Review by derrabus — APPROVED** [review]

**Review by malarzm — APPROVED** [review]
Thanks for this!

File: composer.json [hunk 6]
```diff
@@ -22,17 +22,17 @@
         { "name": "Andreas Braun", "email": "alcaeus@alcaeus.org" }
     ],
     "require": {
-        "php": "^7.2",
+        "php": "^7.2 || ^8.0",
```
Comment (Slamdunk) [inline composer.json:25]: @malarzm can this PR be tagged so we can update to PHP 8?

Comment (malarzm) [reply composer.json:25]: We'll try to have a release in January :)

Comment (Slamdunk) [reply composer.json:25]: No hurry, I'm happy just with a plan :+1:
